crates/runtime/src/dataconnector/graphql.rs

```diff
@@ -79,20 +78,98 @@ Please verify the syntax of your GraphQL query."#
-        Box::pin(async move {
-            let graphql = Self { secret, params };
-            Ok(Arc::new(graphql) as Arc<dyn DataConnector>)
-        })
+impl PaginationParameters {
+    fn parse(query: &str, pointer: &str) -> Option<Self> {
+        let pagination_pattern = r"(?xsm)(\w*)\s*\([^)]*first:\s*(\d+).*\).*\{.*pageInfo.*\{.*hasNextPage.*endCursor.*?\}.*?\}";
+        let regex = unsafe { Regex::new(pagination_pattern).unwrap_unchecked() };
+        match regex.captures(query) {
+            Some(captures) => {
+                let resource_name = captures.get(1).map(|m| m.as_str().to_owned());
+                let count = captures
+                    .get(2)
+                    .map(|m| m.as_str().parse::<usize>())
+                    .transpose()
+                    .unwrap_or(None);
+
+                match (resource_name, count) {
+                    (Some(resource_name), Some(count)) => {
+                        let pattern = format!(r"^(.*?{resource_name})");
+                        let regex = unsafe { Regex::new(pattern.as_str()).unwrap_unchecked() };
+
+                        let captures = regex.captures(pointer);
+
+                        let page_info_path = captures
+                            .and_then(|c| c.get(1).map(|m| m.as_str().to_owned() + "/pageInfo"));
+
+                        page_info_path.map(|page_info_path| Self {
+                            resource_name,
+                            count,
+                            page_info_path,
+                        })
+                    }
+                    _ => None,
+                }
+            }
+            None => None,
+        }
+    }
+
+    fn apply(&self, query: &str, limit: Option<usize>, cursor: Option<String>) -> (String, bool) {
+        let mut limit_reached = false;
+
+        match cursor {
+            Some(cursor) => {
+                let mut count = self.count;
+
+                if let Some(limit) = limit {
+                    if limit < count {
+                        count = limit;
+                        limit_reached = true;
+                    }
+                }
+                let pattern = format!(r#"{}\s*\(.*\)"#, self.resource_name);
+                let regex = unsafe { Regex::new(&pattern).unwrap_unchecked() };
+
+                let new_query = regex.replace(
+                    query,
+                    format!(
+                        r#"{} (first: {count}, after: "{cursor}")"#,
+                        self.resource_name
+                    )
+                    .as_str(),
+                );
+                (new_query.to_string(), limit_reached)
+            }
+            _ => (query.to_string(), false),
+        }
+    }
+
+    fn get_next_cursor_from_response(
+        &self,
+        response: &Value,
+        limit_reached: bool,
+    ) -> Option<String> {
+        if limit_reached {
+            return None;
+        }
```

Comment:
That feels a bit weird - if limit reached I'd expect that `get_next_cursor_from_response` not called at all. Can we make it pure from the response, and check for limit in place?

Reply:
Fixed